### PR TITLE
wolfssl: 5.4.0 -> 5.5.0

### PR DESCRIPTION
**Pull request by LeSuisse**
###### Description of changes

https://github.com/wolfSSL/wolfssl/releases/tag/v5.5.0-stable
Fixes CVE-2022-38152.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>aqemu</li>
    <li>linuxKernel.packages.linux_testing_bcachefs.virtualbox</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>ocamlPackages.ocaml-freestanding</li>
  </ul>
</details>
<details>
  <summary>54 packages built:</summary>
  <ul>
    <li>alpine-make-vm-image</li>
    <li>cloud-init</li>
    <li>cloud-utils</li>
    <li>colima</li>
    <li>cot (python310Packages.cot)</li>
    <li>diffoscope</li>
    <li>gnome.gnome-boxes</li>
    <li>libguestfs</li>
    <li>libguestfs-with-appliance</li>
    <li>lima</li>
    <li>linuxKernel.packages.linux_4_14.virtualbox</li>
    <li>linuxKernel.packages.linux_4_14_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_4_19.virtualbox</li>
    <li>linuxKernel.packages.linux_4_19_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_4_9.virtualbox</li>
    <li>linuxKernel.packages.linux_5_10.virtualbox</li>
    <li>linuxKernel.packages.linux_5_10_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_15.virtualbox</li>
    <li>linuxKernel.packages.linux_hardened.virtualbox (linuxKernel.packages.linux_5_15_hardened.virtualbox)</li>
    <li>linuxKernel.packages.linux_5_18.virtualbox</li>
    <li>linuxKernel.packages.linux_5_18_hardened.virtualbox</li>
    <li>linuxKernel.packages.linux_5_19.virtualbox</li>
    <li>linuxKernel.packages.linux_5_4.virtualbox</li>
    <li>linuxKernel.packages.linux_latest_libre.virtualbox</li>
    <li>linuxKernel.packages.linux_libre.virtualbox</li>
    <li>linuxKernel.packages.linux_lqx.virtualbox</li>
    <li>linuxKernel.packages.linux_xanmod.virtualbox</li>
    <li>linuxKernel.packages.linux_xanmod_latest.virtualbox</li>
    <li>linuxKernel.packages.linux_xanmod_tt.virtualbox</li>
    <li>linuxKernel.packages.linux_zen.virtualbox</li>
    <li>open-watcom-bin</li>
    <li>open-watcom-bin-unwrapped</li>
    <li>out-of-tree</li>
    <li>python310Packages.guestfs</li>
    <li>python39Packages.cot</li>
    <li>python39Packages.guestfs</li>
    <li>qemu</li>
    <li>qemu-utils</li>
    <li>qemu_full</li>
    <li>qemu_kvm</li>
    <li>qemu_test</li>
    <li>qtemu</li>
    <li>quickemu</li>
    <li>rpcs3</li>
    <li>simh</li>
    <li>solo5</li>
    <li>vagrant</li>
    <li>vde2</li>
    <li>virtualbox</li>
    <li>virtualboxHardened</li>
    <li>virtualboxHeadless</li>
    <li>virtualboxWithExtpack</li>
    <li>wolfssl</li>
    <li>zpool-auto-expand-partitions</li>
  </ul>
</details>

`ocamlPackages.ocaml-freestanding` appears to be already broken on `master`.

### CODE REVIEW

**Review by fabaff — COMMENTED** [review]
Looks like that the hash was not updated.

**Comment by LeSuisse** [comment]
It looks like I did not commit the hash modification...

**Comment by LeSuisse** [comment]
For information I took a look at backporting the fix (https://github.com/wolfSSL/wolfssl/commit/91a49da4b8e27be92c57aa9b3c8c3dfc7c5bf6db) to 22.05 but it looks like the patch does not apply cleanly.


```
applying patch /nix/store/h0y1dmmici8pxg4vnwlcr9pp4mxqpppk-CVE-2022-38152.patch
patching file src/internal.c
Hunk #1 succeeded at 6224 (offset -271 lines).
Hunk #2 FAILED at 6867.
Hunk #3 succeeded at 6632 (offset -294 lines).
1 out of 3 hunks FAILED -- saving rejects to file src/internal.c.rej
patching file src/ssl.c
Hunk #1 succeeded at 1734 (offset -45 lines).
Hunk #2 succeeded at 11764 with fuzz 2 (offset -441 lines).
Hunk #3 succeeded at 11797 with fuzz 2 (offset -441 lines).
Hunk #4 FAILED at 12282.
Hunk #5 succeeded at 12210 with fuzz 2 (offset -508 lines).
Hunk #6 succeeded at 17990 with fuzz 1 (offset -630 lines).
Hunk #7 succeeded at 18003 with fuzz 2 (offset -630 lines).
1 out of 7 hunks FAILED -- saving rejects to file src/ssl.c.rej
patching file src/tls.c
Hunk #1 succeeded at 1375 (offset -31 lines).
Hunk #2 FAILED at 5989.
Hunk #3 FAILED at 6009.
2 out of 3 hunks FAILED -- saving rejects to file src/tls.c.rej
patching file src/tls13.c
Hunk #1 succeeded at 8604 with fuzz 1 (offset -1530 lines).
Hunk #2 succeeded at 8616 with fuzz 2 (offset -1531 lines).
Hunk #3 FAILED at 11194.
Hunk #4 succeeded at 9533 with fuzz 2 (offset -1676 lines).
1 out of 4 hunks FAILED -- saving rejects to file src/tls13.c.rej
patching file wolfssl/internal.h
Hunk #1 succeeded at 4675 (offset -387 lines).
error: builder for '/nix/store/7wv8944bz9jpxpw93ykvvfllap8xb8v2-wolfssl-5.3.0.drv' failed with exit code 1;
       last 10 log lines:
       > Hunk #3 FAILED at 6009.
       > 2 out of 3 hunks FAILED -- saving rejects to file src/tls.c.rej
       > patching file src/tls13.c
       > Hunk #1 succeeded at 8604 with fuzz 1 (offset -1530 lines).
       > Hunk #2 succeeded at 8616 with fuzz 2 (offset -1531 lines).
       > Hunk #3 FAILED at 11194.
       > Hunk #4 succeeded at 9533 with fuzz 2 (offset -1676 lines).
       > 1 out of 4 hunks FAILED -- saving rejects to file src/tls13.c.rej
       > patching file wolfssl/internal.h
       > Hunk #1 succeeded at 4675 (offset -387 lines).
       For full logs, run 'nix log /nix/store/7wv8944bz9jpxpw93ykvvfllap8xb8v2-wolfssl-5.3.0.drv'.
```

**Comment by risicle** [comment]
I can take a look after I'm done with the rizen backport.

**Comment by risicle** [comment]
Also CVE-2022-38153

**Comment by risicle** [comment]
Hmm.. I've managed to get the patch for CVE-2022-38152 applying happily & tests passing with a few adjustments.

But CVE-2022-38153 is a different story. Big, complex patch.

**Comment by risicle** [comment]
@ofborg build pkgsMusl.wolfssl

Previously passed, would be a pity to regress.

**Comment by risicle** [comment]
@ofborg eval

(kernel 5.18 removed since last eval, causing errors)

**Comment by risicle** [comment]
Adding to `postPatch`

```
  '' + lib.optionalString stdenv.hostPlatform.isMusl ''
    substituteInPlace scripts/ocsp-stapling2.test \
      --replace '"linux-gnu"' '"linux-musl"'
  '';
```

fixes the tests on musl.

**Comment by LeSuisse** [comment]
Thanks! I was looking at it. Updating the PR.

**Comment by risicle** [comment]
:confused: would have thought that would also catch the `pkgsStatic` case but it doesn't appear to

Perhaps we should just unconditionally change it to `"linux-"', which seems to catch all cases because the code in context uses a wildcard after that.

File: pkgs/development/libraries/wolfssl/default.nix [hunk 1]
```diff
@@ -7,19 +7,21 @@
...
     rev = "v${version}-stable";
-    sha256 = "sha256-5a83Mi+S+mASdZ6O2+0I+qulsF6yNUe80a3qZvWmXHw=";
+    sha256 = "sha256-PqxwWPK5eBcS5d6e0CL4uZHybDye1K8pxniKU99YSAE=";
   };
 
   postPatch = ''
     patchShebangs ./scripts
     # ocsp tests require network access
     sed -i -e '/ocsp\.test/d' -e '/ocsp-stapling\.test/d' scripts/include.am
+    substituteInPlace scripts/ocsp-stapling2.test \
```
Comment (risicle) [inline pkgs/development/libraries/wolfssl/default.nix:23]:
```suggestion
    # ensure test detects musl-based systems too
    substituteInPlace scripts/ocsp-stapling2.test \
```

Else someone will come along and remove it thinking it makes no difference.

**Review by risicle — APPROVED** [review]
`nixpkgs-review` reveals no new failures, macos 10.15 & nixos x86_64. `pkgsMusl` `pkgsStatic` & `pkgsCross.aarch64-multiplatform` variants build. (edit: builds, aarch64 nixos too)

File: pkgs/development/libraries/wolfssl/default.nix (hunk 1 above)
Comment (LeSuisse) [reply pkgs/development/libraries/wolfssl/default.nix:23]: :+1:

**Pull request merged**